c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -356,6 +400,42 @@ struct PrivatePool {
   BlockPool small_blocks;
 };
 
+BlockState::BlockState(Block* block)
+    : stream(block->stream),
+      stream_uses(block->stream_uses),
+      size(block->size),
+      ptr(block->ptr),
+      allocated(block->allocated),
+      gc_count(block->gc_count) {
+  TORCH_CHECK(
+      block->event_count == 0,
+      "Events should have synchronized when checkpointing block");
+};
+
+SegmentState::SegmentState(Block* head) {
+  TORCH_INTERNAL_ASSERT(head->prev == nullptr && head->pool != nullptr);
+  is_small = head->pool->is_small;
+
+  for (Block* curr = head; curr != nullptr; curr = curr->next) {
+    blocks.push_back(std::make_shared<BlockState>(curr));
+  }
+
+  for (size_t i = 1; i < blocks.size(); ++i) {
```

Comment:
Do these need to be recorded? It is always just the next/previous block in the list.